x/concentrated-liquidity/incentives_test.go

```diff
@@ -2987,3 +2970,175 @@ func (s *KeeperTestSuite) TestCreateIncentive() {
+		},
+		"claim and forfeit rewards with varying amounts and different denoms": {
+			growthInside:      uptimeHelper.varyingTokensMultiDenom,
+			growthOutside:     uptimeHelper.varyingTokensSingleDenom,
+			forfeitIncentives: true,
+		},
+	}
+	for _, tc := range tests {
+		tc := tc
+		s.Run(tc.name, func() {
+			// Setup test env.
+			s.SetupTest()
+			clPool := s.PrepareConcentratedPool()
+			clKeeper := s.App.ConcentratedLiquidityKeeper
+
+			// Initialize position.
+			err := clKeeper.InitOrUpdatePosition(s.Ctx, validPoolId, defaultSender, DefaultLowerTick, DefaultUpperTick, sdk.OneDec(), s.Ctx.BlockTime(), time.Hour*24*14)
+			s.Require().NoError(err)
+
+			clPool.SetCurrentTick(DefaultCurrTick)
+			if tc.growthOutside != nil {
+				s.addUptimeGrowthOutsideRange(s.Ctx, validPoolId, defaultSender, DefaultCurrTick.Int64(), DefaultLowerTick, DefaultUpperTick, tc.growthOutside)
+			}
+
+			if tc.growthInside != nil {
+				s.addUptimeGrowthInsideRange(s.Ctx, validPoolId, defaultSender, DefaultCurrTick.Int64(), DefaultLowerTick, DefaultUpperTick, tc.growthInside)
+			}
+
+			err = clKeeper.SetPool(s.Ctx, clPool)
+			s.Require().NoError(err)
+
+			// Store initial accum values for comparison later
+			initUptimeAccumValues, err := clKeeper.GetUptimeAccumulatorValues(s.Ctx, validPoolId)
+			s.Require().NoError(err)
+
+			// Get newly created position to pass into sut function
+			position, err := clKeeper.GetPosition(s.Ctx, validPoolId, defaultSender, DefaultLowerTick, DefaultUpperTick, s.Ctx.BlockTime(), time.Hour*24*14)
+
+			// System under test.
+			amountClaimed, err := clKeeper.ClaimAllIncentivesForPosition(s.Ctx, validPoolId, defaultSender, position, DefaultLowerTick, DefaultUpperTick, tc.forfeitIncentives)
+
+			if tc.expectError != nil {
+				s.Require().Error(err)
+				s.Require().ErrorIs(err, tc.expectError)
+				return
+			}
+			s.Require().NoError(err)
+
+			// We expect claimed rewards to be equal to growth inside
+			expectedCoins := sdk.Coins(nil)
+			for _, growthInside := range tc.growthInside {
+				expectedCoins = expectedCoins.Add(sdk.NormalizeCoins(growthInside)...)
+			}
+			s.Require().Equal(expectedCoins, amountClaimed)
+
+			// Ensure that forfeited incentives were properly added to their respective accumulators
+			if tc.forfeitIncentives {
+				newUptimeAccumValues, err := clKeeper.GetUptimeAccumulatorValues(s.Ctx, validPoolId)
+				s.Require().NoError(err)
+
+				// Subtract the initial accum values to get the delta
+				uptimeAccumDeltaValues, err := osmoutils.SubDecCoinArrays(newUptimeAccumValues, initUptimeAccumValues)
+				s.Require().NoError(err)
+
+				// Convert DecCoins to Coins by truncation for comparison
+				normalizedUptimeAccumDelta := sdk.NewCoins()
+				for _, uptimeAccumDelta := range uptimeAccumDeltaValues {
+					normalizedUptimeAccumDelta = normalizedUptimeAccumDelta.Add(sdk.NormalizeCoins(uptimeAccumDelta)...)
+				}
+
+				s.Require().Equal(normalizedUptimeAccumDelta, amountClaimed)
+			}
+		})
+	}
```

Comment:
Something that would make me feel extra comfortable with this is that we check bank balances before and after claim, to ensure that the rewards in fact did not make it to the account.

Reply:
fd86d6a598ae3654fe93e91a2ea331a673237980